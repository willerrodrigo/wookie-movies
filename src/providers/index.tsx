import { QueryClientProvider } from 'react-query';

import { queryClient } from 'services';
import { ThemeProvider } from 'styled-components/native';
import { theme } from 'theme';

type Props = {
  children: React.ReactNode;
};

export const AppProviders = ({ children }: Props) => (
  <ThemeProvider theme={theme}>
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  </ThemeProvider>
);
