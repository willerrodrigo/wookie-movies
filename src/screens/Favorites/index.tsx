import React, { useCallback } from 'react';

import { MovieItem } from 'components';
import { Movie } from 'services';
import { useFavoriteStore } from 'stores';

import * as S from './styles';

export const Favorites: React.FC = () => {
  const { favorites } = useFavoriteStore();

  const renderMovieItem = useCallback(({ item }: { item: Movie }) => {
    return <MovieItem item={item} />;
  }, []);

  return (
    <S.Wrapper>
      <S.MovieList
        data={favorites}
        keyExtractor={(item) => item.id}
        renderItem={renderMovieItem}
        ListHeaderComponent={<S.Title>{`MY\nFAVORITES`}</S.Title>}
      />
    </S.Wrapper>
  );
};
