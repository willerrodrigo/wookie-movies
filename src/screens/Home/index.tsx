import React, { useCallback, useMemo } from 'react';
import { RefetchOptions } from 'react-query';

import { Loader, Error } from 'components';
import { useMoviesQuery } from 'queries';
import type { Category as CategoryType } from 'services';
import { parseMoviesByGenre } from 'utils/dto';

import { Category } from './components';
import * as S from './styles';

export const Home: React.FC = () => {
  const { data, isRefetching, isError, refetch } = useMoviesQuery();
  const parsedMovies = useMemo(() => parseMoviesByGenre(data || []), [data]);

  const onTryAgain = useCallback(
    (options?: RefetchOptions) => {
      refetch(options);
    },
    [refetch]
  );

  const renderSectionHeader = useCallback(
    ({ section }: { section: CategoryType }) => {
      return <Category category={section} />;
    },
    []
  );

  if (isRefetching) {
    return <Loader />;
  }

  if (isError) {
    return <Error onTryAgain={onTryAgain} />;
  }

  return (
    <S.Wrapper>
      <S.CategoryList
        sections={parsedMovies}
        renderSectionHeader={renderSectionHeader}
        renderItem={() => null}
        ListHeaderComponent={<S.Title>{`WOOKIE\nMOVIES`}</S.Title>}
        SectionSeparatorComponent={() => <S.SectionSeparator />}
      />
    </S.Wrapper>
  );
};
