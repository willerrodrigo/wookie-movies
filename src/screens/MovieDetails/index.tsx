import { useNavigation, useRoute } from '@react-navigation/native';
import React from 'react';
import { useMemo } from 'react';
import { useCallback } from 'react';

import { useFavoriteStore } from 'stores';
import { MovieDetailsParams } from 'types/navigation';

import { Rating } from './components';
import * as S from './styles';

export const MovieDetails: React.FC = () => {
  const navigation = useNavigation();
  const route = useRoute();
  const params = route.params as MovieDetailsParams;
  const {
    id,
    backdrop,
    poster,
    title,
    imdb_rating,
    released_on,
    length,
    director,
    cast,
    overview,
    classification
  } = params.movie;
  const { favoriteMovie, favorites } = useFavoriteStore();
  const directors = Array.isArray(director) ? director.join(', ') : director;

  const handleBackNavigation = useCallback(() => {
    navigation.goBack();
  }, [navigation]);

  const handleFavoriteMovie = useCallback(() => {
    favoriteMovie(params.movie);
  }, [favoriteMovie, params.movie]);

  const isAlreadyFavorited = useMemo(() => {
    return favorites.some((favorite) => favorite.id === id);
  }, [favorites, id]);

  return (
    <S.Wrapper>
      <S.BackdropImg source={{ uri: backdrop }}>
        <S.Overlay />

        <S.Header>
          <S.BackButton onPress={handleBackNavigation}>
            <S.BackButtonIcon />
          </S.BackButton>

          <S.FavoriteButton onPress={handleFavoriteMovie}>
            <S.FavoriteIcon isAlreadyFavorited={isAlreadyFavorited} />
          </S.FavoriteButton>
        </S.Header>
      </S.BackdropImg>

      <S.Content>
        <S.SubHeader>
          <S.Poster source={{ uri: poster }} />

          <S.RatingWrapper>
            <S.Title>
              {title} <S.ImdbRating>({imdb_rating})</S.ImdbRating>
            </S.Title>

            <Rating rating={imdb_rating} />
          </S.RatingWrapper>
        </S.SubHeader>

        <S.Metadata>
          {`${new Date(
            released_on
          ).getFullYear()}  |  ${length}  |  ${classification}  |  ${directors}`}
        </S.Metadata>

        <S.Metadata>Cast: {cast.join(', ')}</S.Metadata>

        <S.Metadata>Movie Description: {overview}</S.Metadata>
      </S.Content>
    </S.Wrapper>
  );
};
