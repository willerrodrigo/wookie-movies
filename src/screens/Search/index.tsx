import React, { useCallback, useState } from 'react';
import { Keyboard } from 'react-native';

import { If, Loader, MovieItem } from 'components';
import { useSearchQuery } from 'queries';
import { Movie } from 'services';

import { SearchInput } from './components';
import * as S from './styles';

export const Search: React.FC = () => {
  const [searchTerm, setSearchTerm] = useState('');
  const { data, isLoading } = useSearchQuery(searchTerm);

  const renderMovieItem = useCallback(({ item }: { item: Movie }) => {
    return <MovieItem item={item} />;
  }, []);

  const handleSearchSubmit = useCallback((searchText: string) => {
    setSearchTerm(searchText);
    Keyboard.dismiss();
  }, []);

  return (
    <S.Wrapper>
      <SearchInput onSearchPress={handleSearchSubmit} />

      <If condition={isLoading}>
        <Loader />
      </If>

      <If condition={!isLoading}>
        <S.MovieList
          data={data || []}
          keyExtractor={(item) => item.id}
          renderItem={renderMovieItem}
        />
      </If>
    </S.Wrapper>
  );
};
