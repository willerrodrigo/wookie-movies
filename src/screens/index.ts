export * from './Home';
export * from './Search';
export * from './MovieDetails';
export * from './Favorites';
