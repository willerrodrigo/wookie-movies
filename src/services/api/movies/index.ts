export * from './types';
export { default as MoviesApi } from './movies.api';
