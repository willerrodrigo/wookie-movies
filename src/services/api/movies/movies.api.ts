import { api } from '../api';
import { Movie } from './types';

export default class MoviesApi {
  static fetchAllMovies = async (): Promise<Movie[]> => {
    try {
      const { data } = await api.get('/movies');

      return data.movies;
    } catch (error) {
      console.error(error);
      throw error;
    }
  };

  static searchMovies = async (searchTerm: string): Promise<Movie[]> => {
    try {
      const { data } = await api.get(`/movies?q=${searchTerm}`);

      return data.movies;
    } catch (error) {
      console.error(error);
      throw error;
    }
  };
}
