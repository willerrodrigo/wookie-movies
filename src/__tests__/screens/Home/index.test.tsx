import {
  cleanup,
  render,
  screen,
  waitFor
} from '@testing-library/react-native';

import { mockedMovies } from '__mocks__/movies';
import Config from '__mocks__/react-native-config';
import nock from 'nock';
import { AppProviders } from 'providers';
import { Home } from 'screens';

afterEach(cleanup);

// jest.useFakeTimers({
//   legacyFakeTimers: true
// });

nock(Config.API_URL)
  // .persist()
  // .defaultReplyHeaders({
  //   'access-control-allow-origin': '*'
  // })
  .get('/movies')
  .reply(200, mockedMovies);

test('should render loading at first', () => {
  const { getByAccessibilityHint } = render(
    <AppProviders>
      <Home />
    </AppProviders>
  );

  getByAccessibilityHint('loading');
});

test('should render movies successfully', async () => {
  const { toJSON } = render(
    <AppProviders>
      <Home />
    </AppProviders>
  );
  const { findAllByText, getByText } = screen;

  const firstGenre = mockedMovies[0].genres[0];

  expect(toJSON()).toMatchSnapshot();
  // await findAllByText(firstGenre);
  // await waitFor(() => {
  //   expect(getByText(/Action/i)).toBeTruthy();
  // });
});
