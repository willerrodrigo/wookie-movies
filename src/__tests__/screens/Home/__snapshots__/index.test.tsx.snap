// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render movies successfully 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "justifyContent": "center",
      },
    ]
  }
>
  <ActivityIndicator
    accessibilityHint="loading"
    color="#191015"
    size={36}
  />
</View>
`;
